### PR TITLE
MAILBOX-374 Enable RabbitMQ EventBus to recover from errors

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java [hunk 1]
```diff
@@ -300,6 +304,40 @@ void startShouldCreateEventExchange() {
...
+            @Test
+            void dispatchShouldWorkAfterNetworkIssuesForOldRegistration() throws Exception {
+                eventBus.start();
+                MailboxListener listener = newListener();
+                eventBus.register(listener, GROUP_A);
+
+                rabbitMQExtension.getRabbitMQ().pause();
+
+                assertThatThrownBy(() -> eventBus.dispatch(EVENT, NO_KEYS).block())
+                    .isNotNull();
```
Comment (trantienduchn) [inline mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java]: you didn't specify exception thrown?

File: mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java [hunk 2]
```diff
@@ -300,6 +304,40 @@ void startShouldCreateEventExchange() {
...
+
+            @Test
+            void dispatchShouldWorkAfterNetworkIssuesForNewRegistration() throws Exception {
+                eventBus.start();
+                MailboxListener listener = newListener();
+
+                rabbitMQExtension.getRabbitMQ().pause();
+
+                assertThatThrownBy(() -> eventBus.dispatch(EVENT, NO_KEYS).block())
+                    .isNotNull();
```
Comment (trantienduchn) [inline mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java]: idem

File: backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQConnectionFactory.java [hunk 3]
```diff
@@ -18,30 +18,47 @@
...
+            this.wrapped = wrapped;
+            connection = Optional.empty();
+        }
+
+        @Override
+        public Connection call() throws Exception {
+            if (connection.map(Connection::isOpen).orElse(false)) {
+                return connection.get();
+            }
+            Connection newConnection = wrapped.call();
```
Comment (trantienduchn) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQConnectionFactory.java]: `wrapped` looks more like a supplier than a Callable. Dont you think? And I suggest to name it to `connectionSupplier`

File: mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java]: No need: I expect that exception

File: backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQConnectionFactory.java (hunk 3 above)
Comment (chibenwa) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQConnectionFactory.java]: A callable is a supplier that can throw exception...

Let's review the naming later on.

Here we do have an issue with key registration as well.... Temporary queues have the good idea to disapear upon connection failures...

Comment (trantienduchn) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQConnectionFactory.java]: I think making them to `non auto delete` and deleting queue manually when unregistering key registration could be a choice.

**Comment by chibenwa** [comment]
Works locally... I'm increasing the wait to 10s

File: backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQConnectionFactory.java (hunk 3 above)
Comment (trantienduchn) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQConnectionFactory.java]: > Temporary queues have the good idea to disapear upon connection failures...

can you add a test for key dispatching to describe this issue?

Comment (chibenwa) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitMQConnectionFactory.java]: Added

File: mailbox/api/src/test/java/org/apache/james/mailbox/events/EventBusTestFixture.java [hunk 4]
```diff
@@ -84,8 +84,8 @@ public void event(Event event) {
     MailboxListener.MailboxEvent EVENT = new MailboxListener.MailboxAdded(SESSION_ID, USER, MAILBOX_PATH, TEST_ID, EVENT_ID);
     MailboxListener.MailboxEvent EVENT_2 = new MailboxListener.MailboxAdded(SESSION_ID, USER, MAILBOX_PATH, TEST_ID, EVENT_ID_2);
 
-    int FIVE_SECOND = 5000;
     int ONE_SECOND = 1000;
+    int THIRTY_SECONDS = 30 * ONE_SECOND;
```
Comment (blackheaven) [inline mailbox/api/src/test/java/org/apache/james/mailbox/events/EventBusTestFixture.java]: I think it's time to use `Duration` (joke intended)

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java [hunk 5]
```diff
@@ -78,10 +78,10 @@
     }
 
     void start() {
-        sender.declareQueue(QueueSpecification.queue()
+        sender.declareQueue(QueueSpecification.queue(eventBusId.asString())
             .durable(DURABLE)
-            .exclusive(EXCLUSIVE)
-            .autoDelete(AUTO_DELETE)
+            .exclusive(!EXCLUSIVE)
```
Comment (blackheaven) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java]: Maybe having `exclusive` and `notExclusive` would be better, I'm not fan of `bool`

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java [hunk 6]
```diff
@@ -78,10 +78,10 @@
...
+            .autoDelete(!AUTO_DELETE)
```
Comment (blackheaven) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java]: idem

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java (hunk 5 above)
Comment (chibenwa) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java]: We avoid negative naming cf @mbaechler 

A local boolean operation is considered more explicit...

Comment (blackheaven) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java]: agreed then

File: mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java [hunk 7]
```diff
@@ -300,6 +304,124 @@ void startShouldCreateEventExchange() {
...
+                rabbitMQExtension.getRabbitMQ().restart();
+
+                eventBus.register(listener, GROUP_A);
+
+                eventBus.dispatch(EVENT, NO_KEYS).block();
+                verify(listener, after(THIRTY_SECONDS).times(1)).event(EVENT);
+            }
+
+            @Test
+            void dispatchShouldWorkAfterRestartForOldKeyRegistration() throws Exception {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/RabbitMQEventBusTest.java]: Asynchronous listeners would also ensure this.

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**